src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1080,15 +1080,29 @@ private HttpResponseHeaders CreateResponseHeaders(bool appCompleted)
                 _keepAlive = false;
             }
 
-            // https://tools.ietf.org/html/rfc7230#section-3.3.1
-            // If any transfer coding other than
-            // chunked is applied to a response payload body, the sender MUST either
-            // apply chunked as the final transfer coding or terminate the message
-            // by closing the connection.
-            if (hasTransferEncoding &&
-                HttpHeaders.GetFinalTransferCoding(responseHeaders.HeaderTransferEncoding) != TransferCoding.Chunked)
+            if (hasTransferEncoding)
             {
-                _keepAlive = false;
+                // https://tools.ietf.org/html/rfc7540#section-8.1.2.2
+                // Such intermediaries SHOULD also remove other connection-specific header fields, such as Keep-Alive,
+                // Proxy-Connection, Transfer-Encoding, and Upgrade, even if they are not nominated by the Connection header field.
+                if (_httpVersion > Http.HttpVersion.Http11)
+                {
+                    responseHeaders.ClearTransferEncoding();
+                    hasTransferEncoding = false;
+                }
+                // https://tools.ietf.org/html/rfc7230#section-3.3.1
+                // If any transfer coding other than
+                // chunked is applied to a response payload body, the sender MUST either
+                // apply chunked as the final transfer coding or terminate the message
+                // by closing the connection.
+                else if (HttpHeaders.GetFinalTransferCoding(responseHeaders.HeaderTransferEncoding) != TransferCoding.Chunked)
+                {
+                    _keepAlive = false;
+                }
+                else if (!ServerOptions.DisableAutoChunking)
+                {
+                    _autoChunk = true;
```

Comment:
The simplest behavior would be just to remove user-set "Transfer-Encoding: chunked" headers and let the normal logic to determine whether or not to auto-chunk take over. This could be done exactly the way it is being done for HTTP/2 and HTTP/3 in this version of the PR:

``
responseHeaders.ClearTransferEncoding();
hasTransferEncoding = false;
```

Trying to change CreateResponseHeaders to simulate what would happen if it wasn't set is a recipe for a lot of subtle bugs. For one, it would be better to write "Content-Length: 0" if the app is completed at this point, but this code doesn't allow that to happen. That one is pretty minor, but there could be other issues that I haven't noticed yet or that could be easily introduced in the future by accident.

Reply:
We should also log something at the debug level to indicate we're ignoring the set "Transfer-Encoding: chunked" response header.